packages/langium/src/grammar/validation/validator.ts

```diff
@@ -906,72 +901,79 @@ export class LangiumGrammarValidator {
-            currentMultiplicity *= 2; // note, that the result is not exact (but it is sufficient for the current case)!
-        }
+    private checkOperatorMultiplicitiesForMultiAssignmentsNested(nodes: AstNode[], parentMultiplicity: number, map: Map<string, AssignmentUse>, accept: ValidationAcceptor): boolean {
+        let resultCreatedNewObject = false;
+        // check all given elements
+        for (let i = 0; i < nodes.length; i++) {
+            const currentNode = nodes[i];
+
+            // Tree-Rewrite-Actions are a special case: a new object is created => following assignments are put into the new object
+            if (ast.isAction(currentNode) && currentNode.feature) {
+                // (This does NOT count for unassigned actions, i.e. actions without feature name, since they change only the type of the current object.)
+                const mapForNewObject = new Map();
+                storeAssignmentUse(mapForNewObject, currentNode.feature, 1, currentNode); // remember the special rewriting feature
+                // all following nodes are put into the new object => check their assignments independently
+                this.checkOperatorMultiplicitiesForMultiAssignmentsIndependent(nodes.slice(i + 1), mapForNewObject, accept);
+                resultCreatedNewObject = true;
+                break; // breaks the current loop
+            }
 
-        // assignment
-        if (ast.isAssignment(currentNode)) {
-            storeAssignmentUse(map, currentNode.feature, 1 * currentMultiplicity, currentNode);
-        }
+            // all other elements don't create new objects themselves:
 
-        // Search for assignments in used fragments as well, since their property values are stored in the current object.
-        // But do not search in calls of regular parser rules, since parser rules create new objects.
-        if (ast.isRuleCall(currentNode) && ast.isParserRule(currentNode.rule.ref) && currentNode.rule.ref.fragment) {
-            this.checkAssignmentNumbersForNode(currentNode.rule.ref.definition, currentMultiplicity, map, accept);
-        }
+            // the current element can occur multiple times => its assignments can occur multiple times as well
+            let currentMultiplicity = parentMultiplicity;
+            if (ast.isAbstractElement(currentNode) && isArrayCardinality(currentNode.cardinality)) {
+                currentMultiplicity *= 2; // note that the result is not exact (but it is sufficient for the current use case)!
+            }
 
-        // rewriting actions are a special case for assignments
-        if (ast.isAction(currentNode) && currentNode.feature) {
-            storeAssignmentUse(map, currentNode.feature, 1 * currentMultiplicity, currentNode);
-        }
+            // assignment
+            if (ast.isAssignment(currentNode)) {
+                storeAssignmentUse(map, currentNode.feature, 1 * currentMultiplicity, currentNode);
```

Comment:
what's the rational of this `1 * ...`, which has no actual effect?

Reply:
It is a relict, in former versions I used `currentMultiplicity` multiple times with different multipliers, where `1 * ` was used to make this multiplier explicit. I removed it now.